server/src/main/java/org/opensearch/transport/TcpTransport.java

```diff
@@ -521,16 +521,42 @@ private BoundTransportAddress createBoundTransportAddress(ProfileSettings profil
             throw new BindTransportException("Failed to resolve publish address", e);
         }
 
-        final int publishPort = resolvePublishPort(profileSettings, boundAddresses, publishInetAddress);
+        final int publishPort = resolvePublishPort(profileSettings.publishPort, boundAddresses, publishInetAddress);
+        if (publishPort == -1) {
+            String profileExplanation = profileSettings.isDefaultProfile ? "" : " for profile " + profileSettings.profileName;
+            throw new BindTransportException(
+                "Failed to auto-resolve publish port"
+                    + profileExplanation
+                    + ", multiple bound addresses "
+                    + boundAddresses
+                    + " with distinct ports and none of them matched the publish address ("
+                    + publishInetAddress
+                    + "). "
+                    + "Please specify a unique port by setting "
+                    + TransportSettings.PORT.getKey()
+                    + " or "
+                    + TransportSettings.PUBLISH_PORT.getKey()
+            );
+        }
+
         final TransportAddress publishAddress = new TransportAddress(new InetSocketAddress(publishInetAddress, publishPort));
         return new BoundTransportAddress(transportBoundAddresses, publishAddress);
     }
 
-    // package private for tests
-    static int resolvePublishPort(ProfileSettings profileSettings, List<InetSocketAddress> boundAddresses, InetAddress publishInetAddress) {
-        int publishPort = profileSettings.publishPort;
-
-        // if port not explicitly provided, search for port of address in boundAddresses that matches publishInetAddress
+    /**
+     * Resolve the publishPort for a server provided a list of boundAddresses and a publishInetAddress.
+     * Resolution strategy is as follows:
+     * If a configured port exists resolve to that port.
+     * If a bound address matches the publishInetAddress resolve to that port.
+     * If a bound address is a wildcard address resolve to that port.
+     * If all bound addresses share the same port resolve to that port.
+     *
+     * @param publishPort -1 if no configured publish port exists
+     * @param boundAddresses addresses bound by the server
+     * @param publishInetAddress address published for the server
+     * @return Resolved port. If publishPort is negative and no port can be resolved return publishPort.
+     */
+    public static int resolvePublishPort(int publishPort, List<InetSocketAddress> boundAddresses, InetAddress publishInetAddress) {
```

Comment:
Could we move these methods `resolveXxx` to `Transport` interface (as suggested before)? They seem to be unrelated to any particular transport